src/metaopt/io/database/mongodb.py

```diff
@@ -0,0 +1,130 @@
+except ImportError:
+    from io import StringIO
+
+try:
+    # Python 3.x
+    from urllib.parse import quote_plus
+except ImportError:
+    # Python 2.x
+    from urllib import quote_plus
+
+class MongoDB(AbstractDB):
+    def __init__(self, dbhost='localhost', dbname='metaopt',
+                dbport=27017, username=None, password=None):
+        super(MongoDB, self).__init__(dbhost, dbname, dbport, username, password)
+
+        # set the connection string
+        # this should be mongodb://mongodb0.example.net:27017
+        # according to MongoDB documentation
+        # https://docs.mongodb.com/getting-started/python/client/
+
+        # try to initialize connection
+        # setting this as a function allows to make sure reconnection
+        # is properly done in the case of an unreliable network
+
+        self.conn = self.initiate_connection()
+        self.db = self.conn[self.dbname]
+
+    @property
+    def uri(self):
+        uri = "mongodb://%s:%s@%s:%s" % (quote_plus(self._username),
+                                         quote_plus(self._password),
+                                         self._dbhost, self._dbport)
+        self._uri = uri
+        return self._uri
+
+    def initiate_connection(self):
+        try:
+            conn = MongoClient(self.uri)
+            self.conn = conn
+        except pymongo.errors.ConnectionFailure as e:
+            self.logger.debug("Could not connect to server")
+            raise_from(DatabaseError("Could not connect to server"), e)
+        # return connection string for checking purposes
+        return conn
+
+    def check_connection(self):
+        if self.db:
+            self.logger.debug("Connection Successful!")
+            return True
+        else:
+            self.logger.debug("Please check your connection")
+            return False
+
+    def close_connection(self):
+        try:
+            close = self.conn.close()
+        except pymongo.errors.ConnectionFailure as e:
+            self.logger.debug("Could not close socket connection")
+            raise_from(DatabaseError("Could not close socket connection."), e)
+        return close
+
+    def is_locked(self):
+        try:
+            is_locked = self.conn.is_locked()
+        except pymongo.errors.ConnectionFailure as e:
+            self.logger.debug("Could not determine if connection is locked.")
+            raise_from(DatabaseError("Could not determine if connection is locked."), e)
+        return is_locked
+
+    def unlock(self):
+        try:
+            unlock = self.conn.unlock()
+        except pymongo.errors.ConnectionFailure as e:
+            self.logger.debug("Could not determine unlock due to connection")
+            raise_from(DatabaseError("Could not determine unlock due to connection."), e)
+        return unlock
+
+    def insert(self, collection_name, query, dictionary):
+
+        dbcollection = self.db[collection_name]
+        result = dbcollection.insert(query, dictionary)
```

Comment:
AFAIK there is no such thing as insert(query, dict) in pymongo. It should be insert_many(query, objects) where objects is a list of dictionaries. Also I can't see why we would want to query something when we do an insert. It's not an update.

Reply:
Insert does exists [What's the difference between insert(), insertOne() and insertMany() methods on MongoDB](https://stackoverflow.com/a/36794912/3492631). I've double checked it with a local execution on my local computer with MongoDB too.

Reply:
Although you're right, it's deprecated, I'll do some changes, although `object` does sound ambiguous, I propose `dictionaries`?

Reply:
Oh! I though it did not exist looking at the doc [here](http://api.mongodb.com/python/current/api/pymongo/collection.html?_ga=2.24207676.782214609.1511212041-2115054929.1508380031#pymongo.collection.Collection.insert_one). Still, insert does not take a query as first argument. I don't understand how your tests can pass :thinking: 

I agree objects is not very clear but dictionaries only tells about the container not the content. What about document_dictionaries?